File: sidecar/models/quant.py
"""QuantLib-backed pricing Pydantic models — Phase 6.

Hand-maintained Python mirror of ``types/quant.ts``.

These shapes are framework-neutral on purpose — no QuantLib types appear on
the wire. The QuantLib backend in ``services/quant/`` converts these inputs
into its internal C++ object model and back to plain dataclasses for the
HTTP boundary.
"""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict

# ---------------------------------------------------------------------------
# Input-domain bounds (Phase 9.5 F4-1)
# ---------------------------------------------------------------------------
#
# The quant endpoints previously did ZERO input-domain validation: negative
# volatility, strike 0, expiry<=valuation, and absurd ytm all returned 200 with
# degenerate/garbage output. Each request model now exposes ``validate_domain()``
# raising a clear ``ValueError`` for economically-invalid inputs; the quant
# router already maps ``ValueError`` -> HTTP 400. These bounds are deliberately
# generous (any real-world scenario fits) — they reject only nonsense that would
# otherwise produce garbage prices or a DoS-grade compute load.

MIN_RATE = -1.0  # a yield / rate below -100% is economically meaningless
MAX_RATE = 10.0  # 1000% — far above any real rate, rejects 1e6-style garbage
MAX_VOLATILITY = 100.0  # 10000% annualised vol ceiling
MAX_BINOMIAL_STEPS = 100_000  # guard against runaway lattice compute
MAX_MC_PATHS = 10_000_000  # guard against runaway Monte-Carlo compute
MAX_CURVE_SAMPLES = 10_000  # guard against runaway curve sampling


def _validate_option_inputs(
    *,
    spot: float,
    strike: float,
    volatility: float,
    risk_free_rate: float,
    dividend_yield: float,
    valuation_date: date,
    expiry_date: date,
) -> None:
    """Shared domain checks for option / Greeks requests (raises ValueError)."""
    if spot <= 0:
        raise ValueError(f"spot must be positive (got {spot})")
    if strike <= 0:
        raise ValueError(f"strike must be positive (got {strike})")
    if not 0 <= volatility <= MAX_VOLATILITY:
        raise ValueError(f"volatility must be between 0 and {MAX_VOLATILITY} (got {volatility})")
    if not MIN_RATE <= risk_free_rate <= MAX_RATE:
        raise ValueError(
            f"risk_free_rate must be between {MIN_RATE} and {MAX_RATE} (got {risk_free_rate})"
        )
    if not 0 <= dividend_yield <= MAX_RATE:
        raise ValueError(f"dividend_yield must be between 0 and {MAX_RATE} (got {dividend_yield})")
    if expiry_date <= valuation_date:
        raise ValueError("expiry_date must be after valuation_date")


# ---------------------------------------------------------------------------
# Options
# ---------------------------------------------------------------------------

OptionExercise = Literal["european", "american"]
OptionPayoff = Literal["call", "put"]
OptionPricingMethod = Literal["black-scholes", "binomial", "monte-carlo"]


class OptionPricingRequest(BaseModel):
    """Inputs for an option-pricing call."""

    model_config = ConfigDict(extra="forbid")

    exercise: OptionExercise
    payoff: OptionPayoff
    spot: float
    strike: float
    risk_free_rate: float
    dividend_yield: float
    volatility: float
    valuation_date: date
    expiry_date: date
    method: OptionPricingMethod
    binomial_steps: int | None = None
    monte_carlo_paths: int | None = None
    monte_carlo_seed: int | None = None

    def validate_domain(self) -> None:
        """Reject economically-invalid inputs (surfaced as HTTP 400). F4-1."""
        _validate_option_inputs(
            spot=self.spot,
            strike=self.strike,
            volatility=self.volatility,
            risk_free_rate=self.risk_free_rate,
            dividend_yield=self.dividend_yield,
            valuation_date=self.valuation_date,
            expiry_date=self.expiry_date,
        )
        if self.binomial_steps is not None and not 1 <= self.binomial_steps <= MAX_BINOMIAL_STEPS:
            raise ValueError(
                f"binomial_steps must be between 1 and {MAX_BINOMIAL_STEPS} "
                f"(got {self.binomial_steps})"
            )
        if self.monte_carlo_paths is not None and not 1 <= self.monte_carlo_paths <= MAX_MC_PATHS:
            raise ValueError(
                f"monte_carlo_paths must be between 1 and {MAX_MC_PATHS} "
                f"(got {self.monte_carlo_paths})"
            )


class Greeks(BaseModel):
    """First / second-order option sensitivities."""

    model_config = ConfigDict(extra="forbid")

    delta: float
    gamma: float
    vega: float
    theta: float
    rho: float


class OptionPricingResult(BaseModel):
    """Output of an option-pricing call."""

    model_config = ConfigDict(extra="forbid")

    price: float
    greeks: Greeks | None = None
    method: OptionPricingMethod
    monte_carlo_std_error: float | None = None
    duration_ms: float


# ---------------------------------------------------------------------------
# Bonds
# ---------------------------------------------------------------------------

CouponFrequency = Literal[1, 2, 4]


class BondPricingRequest(BaseModel):
    """Inputs for a fixed-rate bond pricing call."""

    model_config = ConfigDict(extra="forbid")

    face_value: float = 1000.0
    coupon_rate: float
    coupons_per_year: CouponFrequency
    issue_date: date
    maturity_date: date
    settlement_date: date
    yield_to_maturity: float

    def validate_domain(self) -> None:
        """Reject economically-invalid numeric inputs (surfaced as HTTP 400).

        Date ordering (issue/settlement/maturity) is enforced separately in
        ``services.quant.bonds.price_bond`` — this covers only the numeric
        domain the catalog flagged (face/coupon/ytm). F4-1.
        """
        if self.face_value <= 0:
            raise ValueError(f"face_value must be positive (got {self.face_value})")
        if not 0 <= self.coupon_rate <= MAX_RATE:
            raise ValueError(
                f"coupon_rate must be between 0 and {MAX_RATE} (got {self.coupon_rate})"
            )
        if not MIN_RATE < self.yield_to_maturity <= MAX_RATE:
            raise ValueError(
                f"yield_to_maturity must be greater than {MIN_RATE} and at most {MAX_RATE} "
                f"(got {self.yield_to_maturity})"
            )


class BondPricingResult(BaseModel):
    """Output of a bond-pricing call."""

    model_config = ConfigDict(extra="forbid")

    clean_price: float
    dirty_price: float
    accrued_interest: float
    duration: float
    modified_duration: float
    convexity: float
    duration_ms: float


# ---------------------------------------------------------------------------
# Yield curve
# ---------------------------------------------------------------------------

YieldCurveInstrumentType = Literal["deposit", "swap"]
TenorUnit = Literal["months", "years"]


class YieldCurveInstrument(BaseModel):
    """One instrument used to bootstrap a yield curve."""

    model_config = ConfigDict(extra="forbid")

    type: YieldCurveInstrumentType
    tenor: int
    tenor_unit: TenorUnit
    rate: float


class YieldCurveRequest(BaseModel):
    """Bootstrap a yield curve from the given instruments."""

    model_config = ConfigDict(extra="forbid")

    valuation_date: date
    instruments: list[YieldCurveInstrument]
    sample_count: int

    def validate_domain(self) -> None:
        """Reject invalid bootstrap inputs (surfaced as HTTP 400). F4-1."""
        if not self.instruments:
            raise ValueError("yield-curve bootstrap requires at least one instrument")
        if not 2 <= self.sample_count <= MAX_CURVE_SAMPLES:
            raise ValueError(
                f"sample_count must be between 2 and {MAX_CURVE_SAMPLES} (got {self.sample_count})"
            )
        for inst in self.instruments:
            if inst.tenor <= 0:
                raise ValueError(f"instrument tenor must be positive (got {inst.tenor})")
            if not MIN_RATE <= inst.rate <= MAX_RATE:
                raise ValueError(
                    f"instrument rate must be between {MIN_RATE} and {MAX_RATE} (got {inst.rate})"
                )


class YieldCurvePoint(BaseModel):
    """One sampled point on a bootstrapped curve."""

    model_config = ConfigDict(extra="forbid")

    date: date
    tenor_years: float
    zero_rate: float
    discount_factor: float


class YieldCurveResult(BaseModel):
    """Output of ``POST /quant/yield-curve``."""

    model_config = ConfigDict(extra="forbid")

    valuation_date: date
    curve: list[YieldCurvePoint]
    duration_ms: float


# ---------------------------------------------------------------------------
# Greeks-only (dashboard helper)
# ---------------------------------------------------------------------------


class GreeksRequest(BaseModel):
    """Subset of :class:`OptionPricingRequest` for the standalone Greeks endpoint."""

    model_config = ConfigDict(extra="forbid")

    payoff: OptionPayoff
    spot: float
    strike: float
    risk_free_rate: float
    dividend_yield: float
    volatility: float
    valuation_date: date
    expiry_date: date

    def validate_domain(self) -> None:
        """Reject economically-invalid inputs (surfaced as HTTP 400). F4-1."""
        _validate_option_inputs(
            spot=self.spot,
            strike=self.strike,
            volatility=self.volatility,
            risk_free_rate=self.risk_free_rate,
            dividend_yield=self.dividend_yield,
            valuation_date=self.valuation_date,
            expiry_date=self.expiry_date,
        )


class GreeksResult(BaseModel):
    """Output of ``POST /quant/option/greeks``."""

    model_config = ConfigDict(extra="forbid")

    greeks: Greeks
    price: float
    duration_ms: float

File: types/quant.ts
/**
 * Vysted Terminal — QuantLib-backed pricing contracts (Phase 6).
 *
 * Hand-maintained TypeScript mirror of ``sidecar/models/quant.py``.
 *
 * All pricing is in-process via the QuantLib==1.42.1 Python wrapper bundled
 * into the main sidecar (Tier-3 architecture choice — see
 * ``docs/PHASE_6_HANDOFF.md`` and CHANGELOG v0.6.0). The wire shapes here
 * are deliberately framework-neutral — no QuantLib types leak into the
 * frontend.
 */

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

/** European or American exercise. v0.6.0 ships both. */
export type OptionExercise = "european" | "american";

/** Call or put. */
export type OptionPayoff = "call" | "put";

/**
 * Which numerical engine to use. ``black-scholes`` is closed-form European;
 * ``binomial`` is Cox-Ross-Rubinstein American; ``monte-carlo`` covers
 * path-dependent payoffs the Asian / Barrier engines also support.
 */
export type OptionPricingMethod = "black-scholes" | "binomial" | "monte-carlo";

/**
 * Inputs for an option-pricing call. Day-count + calendar are fixed
 * server-side (``Actual365Fixed`` + ``NullCalendar``) — the wire stays
 * minimal; advanced users get a custom-config endpoint later if needed.
 */
export interface OptionPricingRequest {
  exercise: OptionExercise;
  payoff: OptionPayoff;
  /** Spot price of the underlying at valuation time. */
  spot: number;
  /** Strike price. */
  strike: number;
  /** Risk-free rate, continuously compounded annualised (e.g. 0.05 for 5 %). */
  risk_free_rate: number;
  /** Continuous dividend yield (0.0 for non-dividend-paying). */
  dividend_yield: number;
  /** Annualised volatility (e.g. 0.30 for 30 %). */
  volatility: number;
  /** Valuation date (ISO-8601). */
  valuation_date: string;
  /** Expiry date (ISO-8601). Must be > valuation_date. */
  expiry_date: string;
  method: OptionPricingMethod;
  /** Binomial: tree steps (default 200). MC: ignored. */
  binomial_steps?: number;
  /** MC: number of sample paths (default 50000). Binomial/BS: ignored. */
  monte_carlo_paths?: number;
  /** MC: random seed for reproducible runs. */
  monte_carlo_seed?: number;
}

/**
 * The Greeks — first / second-order sensitivities. All values are per-unit
 * (delta in shares-per-option, gamma in per-spot, vega in per-vol-point /100,
 * theta in per-calendar-day, rho in per-rate-point /100). The Vysted UI
 * relabels into the standard display conventions before showing.
 */
export interface Greeks {
  delta: number;
  gamma: number;
  vega: number;
  theta: number;
  rho: number;
}

/**
 * Output of an option-pricing call. Greeks are populated when the chosen
 * method supports them (Black-Scholes always; binomial via finite-difference;
 * Monte Carlo via pathwise derivative — ``null`` when not computed).
 */
export interface OptionPricingResult {
  price: number;
  greeks: Greeks | null;
  method: OptionPricingMethod;
  /** Standard error of the MC estimate (Monte Carlo only). */
  monte_carlo_std_error: number | null;
  /** Computation time on the sidecar (ms). */
  duration_ms: number;
}

// ---------------------------------------------------------------------------
// Bonds
// ---------------------------------------------------------------------------

/**
 * Inputs for a fixed-rate bond pricing call. Supports annual / semi-annual
 * / quarterly coupons. Day-count is ``Thirty360`` per the US convention;
 * compounding is ``Compounded`` and frequency follows the coupon frequency.
 */
export interface BondPricingRequest {
  /** Bond face value (default 1000.0). */
  face_value: number;
  /** Annual coupon rate (e.g. 0.05 for 5 %). */
  coupon_rate: number;
  /** Coupons per year — 1 (annual), 2 (semi-annual), 4 (quarterly). */
  coupons_per_year: 1 | 2 | 4;
  /** Issue date (ISO-8601). */
  issue_date: string;
  /** Maturity date (ISO-8601). */
  maturity_date: string;
  /** Settlement date (ISO-8601). Used as the valuation date. */
  settlement_date: string;
  /** Yield-to-maturity at which to price (annualised, decimal). */
  yield_to_maturity: number;
}

/** Output of a bond-pricing call. */
export interface BondPricingResult {
  /** Clean price (per 100 face). */
  clean_price: number;
  /** Dirty price (clean + accrued). */
  dirty_price: number;
  accrued_interest: number;
  /** Macaulay duration. */
  duration: number;
  /** Modified duration. */
  modified_duration: number;
  convexity: number;
  duration_ms: number;
}

// ---------------------------------------------------------------------------
// Yield curve
// ---------------------------------------------------------------------------

/**
 * One instrument used to bootstrap a yield curve. v0.6.0 supports ``deposit``
 * (money-market rate) + ``swap`` (interest-rate swap). The curve panel
 * lets the user paste a tenor-rate grid and bootstrap directly.
 */
export interface YieldCurveInstrument {
  /** Instrument type. */
  type: "deposit" | "swap";
  /** Tenor in months for deposits (e.g. 1, 3, 6) or years for swaps
   * (e.g. 2, 5, 10, 30) — the wire just carries the numeric tenor with a
   * ``unit`` field. */
  tenor: number;
  tenor_unit: "months" | "years";
  /** Rate (annualised, decimal). */
  rate: number;
}

/**
 * Bootstrap a yield curve from the given instruments + valuation date. The
 * curve is sampled at evenly-spaced node dates for the response — the panel
 * renders this as a line chart.
 */
export interface YieldCurveRequest {
  valuation_date: string;
  instruments: YieldCurveInstrument[];
  /** How many sample points to return (the bootstrapped curve is then
   * interpolated to those dates). */
  sample_count: number;
}

/** One sampled point on a bootstrapped curve. */
export interface YieldCurvePoint {
  /** ISO-8601 sample date. */
  date: string;
  /** Tenor in years (decimal). */
  tenor_years: number;
  /** Zero-rate at that tenor (continuously compounded, decimal). */
  zero_rate: number;
  /** Discount factor at that tenor. */
  discount_factor: number;
}

export interface YieldCurveResult {
  valuation_date: string;
  curve: YieldCurvePoint[];
  duration_ms: number;
}

// ---------------------------------------------------------------------------
// Greeks-only request (for the Greeks dashboard)
// ---------------------------------------------------------------------------

/** Subset of ``OptionPricingRequest`` for the standalone Greeks endpoint —
 * always uses Black-Scholes analytic engine. */
export interface GreeksRequest {
  payoff: OptionPayoff;
  spot: number;
  strike: number;
  risk_free_rate: number;
  dividend_yield: number;
  volatility: number;
  valuation_date: string;
  expiry_date: string;
}

export interface GreeksResult {
  greeks: Greeks;
  /** Black-Scholes price returned alongside the Greeks for free. */
  price: number;
  duration_ms: number;
}
